docs/settings.rst
========
Settings
========

Waliki follows a `convention over configuration <http://en.wikipedia.org/wiki/Convention_over_configuration>`_
paradigm, defining sensible defaults for every constant.

You can override any settings in your project's :file:`settings.py` file


.. confval:: WALIKI_DATA_DIR

    Waliki's content path. By default it's ``<project_root>/waliki_data``

    You can set it to an absolute path. Ensure the path exists and it's writable from
    your web server

.. confval:: WALIKI_INDEX_SLUG

   The slug of the index page. Default is ``home``


.. confval:: WALIKI_ANONYMOUS_USER_PERMISSIONS

    The tuple of permissions given to not authenticated users. Default is ``('view_page',)``
    Check :ref:`acl` for further details. If there is no ``change_page`` permission,
    the anonymous user is redirected to the login page when try to edit the page.

.. confval:: WALIKI_LOGGED_USER_PERMISSIONS

    The tuple of permissions given to any authenticated user. Default is ``('view_page', 'add_page', 'change_page')``. Check :ref:`acl` for further details.



.. confval:: WALIKI_AVAILABLE_MARKUPS

    A list tha define the enabled markups. Default is ``['reStructuredText', 'Markdown']``.
    Available markups are ``reStructuredText``, ``Markdown`` and ``Textile``

.. confval:: WALIKI_DEFAULT_MARKUP

    The default markup for new pages. Default ``WALIKI_AVAILABLE_MARKUPS[0]``

.. confval:: WALIKI_SLUG_PATTERN

    Pattern used in urls to match any page related view, wich is also de filename of
    the file that store the page content. Default is ``'[a-zA-Z0-9-_\/]+'``.


.. confval:: WALIKI_SLUGIFY_FUNCTION

    String pointing to a callable that receive a text and return and slug.
    Default ``'waliki.utils.get_slug'``

    If you override it, ensure that ``your_get_slug(any_valid_slug) == any_valid_slug``


.. confval:: WALIKI_SANITIZE_FUNCTION

    .. versionadded:: 0.6

    String pointing to a callable that receive html and return and return a sanitized version of it.
    Default ``'waliki.utils.sanitize'``, which just removes ``<script>`` tags.

    You can define a more sofisticated version using `bleach <http://bleach.readthedocs.org>`_ or
    lxml's `Cleaner <http://lxml.de/api/lxml.html.clean.Cleaner-class.html>`_


.. confval:: WALIKI_MARKUPS_SETTINGS

    Dictionary of keywords arguments to extend or override the ones passed for each markup class.
    By default, this is the dictionary used

    .. code-block:: python

        {'reStructuredText': {
            # check http://docutils.sourceforge.net/docs/user/config.html
            'settings_overrides': {
                'initial_header_level': 2,
                'record_dependencies': True,
                'stylesheet_path': None,
                'link_stylesheet': True,
                'syntax_highlight': 'short',
                'halt_level': 5,
            },
            'writer': WalikiHTML5Writer(),
            'writer_name': 'html5',
            },
        'Markdown': {
                'extensions': ['wikilinks', 'headerid'],
                'extension_configs': {
                    'wikilinks': {'build_url': get_url},
                    'headerid': {'level': 2},
                }
            }
        }


.. confval:: WALIKI_BREADCRUMBS

    .. versionadded:: 0.6

    If ``True``, show a breadcrumbs with links to "parent" pages. Default is ``False``


.. confval:: WALIKI_PDF_INCLUDE_TITLE

    Apply if :ref:`PDF plugin <pdf>` is installed.

    As the title is not part of the file content but stored in the database, it should be given
    to rst2pdf. Default is ``False``

.. confval:: WALIKI_PDF_RST2PDF_BIN

    Apply if :ref:`PDF plugin <pdf>` is installed.

    A custom binary path to rst2pdf. E.g. '/usr/bin/rst2pdf'

.. confval:: WALIKI_CODEMIRROR_SETTINGS

    A dictionary (converted to json) used to `configure Codemirror <http://codemirror.net/doc/manual.html#config>`_. The default is:

    .. code-block:: python

        {'theme': 'mbo', 'autofocus': True, 'lineNumbers': True}

.. confval:: WALIKI_RENDER_403

   If ``True``, raise an HTTP 403 (Forbidden error) if an authenticated user is not allowed to edit a page. Default is ``True``.

.. confval:: WALIKI_PAGINATE_BY

   The numbers of items per page in paginated lists, for example "what changed". Default is ``20``.

.. confval:: WALIKI_COMMITTER_EMAIL

    If :ref:`git` is enabled and anonymous editios allowed, this is the git's committer email used. Default is ``waliki@waliki.pythonanywhere.com``.


.. confval:: WALIKI_COMMITTER_NAME

    Analog to :confval:`WALIKI_COMMITTER_EMAIL`. Default is ``Waliki``

.. confval:: WALIKI_CACHE_TIMEOUT

    The maximum expiration time for a page cache, in seconds. Default is ``60*60*24`` (i.e. 1 day)

.. confval:: WALIKI_ATTACHMENTS_DIR

   If :ref:`attachments` is enabled, this is the path where uploaded files are stored.

   By default it's ``<project_root>/waliki_attachments``. Ensure the path exists and it's writable by your web server.

.. confval:: WALIKI_UPLOAD_TO_PATTERN

   The pattern used in the path relative to :confval:`WALIKI_ATTACHMENTS_DIR` to store uploaded files. It's interpolated with the following dictionary:

   .. code-block:: python

        {'slug': instance.page.slug,
        'page_id': getattr(instance.page, 'id', ''),
        'filename': filename,
        'filename_extension': os.path.splitext(filename)[1]}

    Default is ``'%(slug)s/%(filename)s'``

.. confval:: WALIKI_RST_DIRECTIVES

    List of string poiting to modules with ``register_directive()`` function that register
    extra reStructuredText Directives. Default is ``['waliki.directives.embed']``

    Check `embed.py <https://github.com/mgaitan/waliki/blob/master/waliki/directives/embed.py>`_  as an example.

.. confval:: WALIKI_RST_TRANSFORMS

    List of string poiting to reStructuredText extra Transforms classes to be applied

    Check `transforms.py <https://github.com/mgaitan/waliki/blob/master/waliki/directives/transforms.py>`_  as an example.

    Default is ``['waliki.directives.transforms.Emojis']``

.. confval:: WALIKI_USE_MATHJAX

    If ``True``, load Mathjax's assets from the official CDN service
    Default is ``False``. Check the :ref:`faq <math>` for details.



HISTORY.rst
.. :changelog:

Changelog
---------

0.9 (under dev)
++++++++++++++++++

- Added compatibility with Django 1.11

0.8.1 (2017-03-26)
++++++++++++++++++

- Fixed compatibiltiy with Django 1.10 (thangs to `Martí Bosch`_)
- Fixed `#125 <https://github.com/mgaitan/waliki/issues/125>`__
- Upgraded demo's setting to recent format
- Added missing migration

.. _Martí Bosch: https://github.com/martibosch

0.7 (2016-12-19)
++++++++++++++++

- Fix compatibility with Django 1.9.x and Markup 2.x (thanks to `Oleg Girko`_ for the contribution)

.. _Oleg Girko: https://github.com/OlegGirko


0.6 (2015-10-25)
+++++++++++++++++

- Slides view use the cache. Fix `#81 <https://github.com/mgaitan/waliki/issues/81>`__
- Implemented an RSS feed listing lastest changes. It's part of `#32 <https://github.com/mgaitan/waliki/issues/32>`__
- Added a `configurable "sanitize" <http://waliki.readthedocs.org/en/latest/settings.html#confval-WALIKI_SANITIZE_FUNCTION>`_ function.
- Links to attachments doesn't relay on IDs by default (but it's backaward compatible).  `#96 <https://github.com/mgaitan/waliki/issues/32>`_
- Added an optional "`breadcrumb <http://waliki.readthedocs.org/en/latest/settings.html#confval-WALIKI_BREADCRUMBS>`_ " hierarchical links for pages. `#110 <https://github.com/mgaitan/waliki/pull/110>`_
- Run git with output to pipe instead of virtual terminal. `#111 <https://github.com/mgaitan/waliki/pull/111>`_

0.5 (2015-04-12)
++++++++++++++++++

- Per page markup is now fully functional. It allows to
  have a mixed rst & markdown wiki. Fixed `#2 <https://github.com/mgaitan/waliki/issues/2>`__
- Allow save a page without changes in a body.
  Fixed `#85 <https://github.com/mgaitan/waliki/issues/85>`__
- Fixed `#84 <https://github.com/mgaitan/waliki/issues/84>`__, that marked deleted but no commited after a move
- Allow to choice markup from new page dialog. `#82 <https://github.com/mgaitan/waliki/issues/82>`__
- Fix wrong encoding for raw of an old revision. `#75 <https://github.com/mgaitan/waliki/issues/75>`__


0.4.2 (2015-03-31)
++++++++++++++++++

- Fixed conflict with a broken dependecy


0.4.1 (2015-03-31)
++++++++++++++++++

- Marked the release as beta (instead of alpha)
- Improves on setup.py and the README

0.4 (2015-03-31)
++++++++++++++++

- Implemented views to add a new, move and delete pages
- Implemented real-time collaborative editing via together.js
  (`#33 <https://github.com/mgaitan/waliki/issues/33>`__)
- Added pagination in *what changed* page
- Added a way to extend waliki's docutils with directives and transformation for
- A deep docs proofreading by `chuna <https://github.com/chuna>`__
- Edit view redirect to detail if the page doesn't exist
  (`#37 <https://github.com/mgaitan/waliki/issues/37>`__)
- waliki\_box fails with missing slug
  `#40 <https://github.com/mgaitan/waliki/issues/40>`__
- can't view diffs on LMDE
  `#60 <https://github.com/mgaitan/waliki/issues/60>`__
- fix typos in tutorial
  `#76 <https://github.com/mgaitan/waliki/pull/76>`__
  (`martenson <https://github.com/martenson>`__)
- Fix build with Markups 0.6.
  `#63 <https://github.com/mgaitan/waliki/pull/63>`__
  (`loganchien <https://github.com/loganchien>`__)
- fixed roundoff error for whatchanged pagination
  `#61 <https://github.com/mgaitan/waliki/pull/61>`__
  (`aszepieniec <https://github.com/aszepieniec>`__)

- Enhance slides `#59 <https://github.com/mgaitan/waliki/pull/59>`__
  (`loganchien <https://github.com/loganchien>`__)

- Fix UnicodeDecodeError in waliki.git.view.
  `#58 <https://github.com/mgaitan/waliki/pull/58>`__
  (`loganchien <https://github.com/loganchien>`__)

0.3.3 (2014-11-24)
++++++++++++++++++

- Tracking page redirections
- fix bugs related to attachments in `sync_waliki`
- The edition form uses crispy forms if it's installed
- many small improvements to help the integration/customization

0.3.2 (2014-11-17)
++++++++++++++++++

- Url pattern is configurable now. By default allow uppercase and underscores
- Added ``moin_migration_cleanup``, a tool to cleanup the result of a moin2git_ import
- Improve git parsers for *page history* and *what changed*

.. _moin2git: https://github.com/mgaitan/moin2git


0.3.1 (2014-11-11)
++++++++++++++++++

- Plugin *attachments*
- Implemented *per namespace* ACL rules
- Added the ``waliki_box`` templatetag: use waliki content in any app
- Added ``entry_point`` to extend templates from plugins
- Added a webhook to pull and sync change from a remote repository (Git)
- Fixed a bug in git that left the repo unclean

0.2 (2014-09-29)
++++++++++++++++

- Support concurrent edition
- Added a simple ACL system
- ``i18n`` support (and locales for ``es``)
- Editor based in Codemirror
- Migrated templates to Bootstrap 3
- Added the management command ``waliki_sync``
- Added a basic test suite and setup Travis CI.
- Added "What changed" page (from Git)
- Plugins can register links in the nabvar (``{% navbar_links %}``)

0.1.2 / 0.1.3 (2014-10-02)
++++++++++++++++++++++++++

* "Get as PDF" plugin
* rst2html5 fixes

0.1.1 (2014-10-02)
++++++++++++++++++

* Many Python 2/3 compatibility fixes

0.1.0 (2014-10-01)
++++++++++++++++++

* First release on PyPI.

waliki/settings.py
import os.path
from importlib import import_module
import collections
from django.conf import settings
from .utils import get_url
from .directives.writer import WalikiHTML5Writer
try:
    from django.utils.module_loading import import_string
except ImportError:
    # django 1.6
    from django.utils.module_loading import import_by_path as import_string


def _get_default_data_dir(directory, abspath=True):
    settings_mod = import_module(settings.SETTINGS_MODULE)

    project_dir = os.path.dirname(settings_mod.__name__)
    if abspath:
        project_dir = os.path.abspath(project_dir)
    return os.path.join(project_dir, directory)


def deep_update(d, u):
    """update nested dicts. if u is a dict and hasn't a key, the original is keeped
    inspired in Alex Martelli's
    http://stackoverflow.com/a/3233356"""
    for k, v in u.items():
        if isinstance(v, collections.Mapping):
            r = deep_update(d.get(k, {}), v)
            d[k] = r
        else:
            d[k] = u[k]
    return d


def _get_markup_settings(user_settings):
    defaults = {'reStructuredText': {
                    # check http://docutils.sourceforge.net/docs/user/config.html
                    'settings_overrides': {              # noqa
                        'initial_header_level': 2,
                        'record_dependencies': True,
                        'stylesheet_path': None,
                        'link_stylesheet': True,
                        'syntax_highlight': 'short',
                        'halt_level': 5,
                    },
                    'writer': WalikiHTML5Writer(),
                    'writer_name': 'html5',
                    },
                'Markdown': {
                    'extensions': ['wikilinks', 'toc'],
                    'extension_configs': {
                        'wikilinks': {'build_url': get_url},
                        'toc': {'baselevel': 2},
                    }
                }
            }

    deep_update(defaults, user_settings)
    return defaults

WALIKI_AVAILABLE_MARKUPS = getattr(settings, 'WALIKI_AVAILABLE_MARKUPS', ['reStructuredText', 'Markdown'])

# options: reStructuredText, Markdown, Textile
WALIKI_DEFAULT_MARKUP = WALIKI_AVAILABLE_MARKUPS[0]


WALIKI_SLUG_PATTERN = getattr(settings, 'WALIKI_SLUG_PATTERN', '[a-zA-Z0-9-_\/]+')

WALIKI_SLUGIFY_FUNCTION = getattr(settings, 'WALIKI_SLUGIFY_FUNCTION', 'waliki.utils.get_slug')

WALIKI_SANITIZE_FUNCTION = getattr(settings, 'WALIKI_SANITIZE_FUNCTION', 'waliki.utils.sanitize')

get_slug = import_string(WALIKI_SLUGIFY_FUNCTION)
sanitize = import_string(WALIKI_SANITIZE_FUNCTION)


# your content folder. by default it's <project_root>/waliki_data
WALIKI_DATA_DIR = getattr(settings, 'WALIKI_DATA_DIR', None) or _get_default_data_dir('waliki_data')

# wich page is shown as the wiki index?
WALIKI_INDEX_SLUG = getattr(settings, 'WALIKI_INDEX_SLUG', "home")

# kwargs for each markup.
WALIKI_MARKUPS_SETTINGS = _get_markup_settings(getattr(settings, 'WALIKI_MARKUPS_SETTINGS', {}))

# get as txt
WALIKI_PDF_INCLUDE_TITLE = getattr(settings, 'WALIKI_PDF_INCLUDE_TITLE', False)

# custom rst2pdf binary path. You should set it on python3.
WALIKI_PDF_RST2PDF_BIN = getattr(settings, 'WALIKI_PDF_RST2PDF_BIN', False)

WALIKI_CODEMIRROR_SETTINGS = getattr(settings, 'WALIKI_CODEMIRROR_SETTINGS',
                                     {'lineNumbers': False, 'theme': 'mbo', 'autofocus': True, 'lineNumbers': True})

# ('view_page', 'add_page', 'change_page', 'delete_page')
WALIKI_ANONYMOUS_USER_PERMISSIONS = getattr(settings, 'WALIKI_ANONYMOUS_USER_PERMISSIONS',
                                            ('view_page',))

WALIKI_LOGGED_USER_PERMISSIONS = getattr(settings, 'WALIKI_LOGGED_USER_PERMISSIONS',
                                         ('view_page', 'add_page', 'change_page'))

WALIKI_RENDER_403 = getattr(settings, 'WALIKI_RENDER_403', True)

WALIKI_PAGINATE_BY = getattr(settings, 'WALIKI_PAGINATE_BY', 20)

WALIKI_COMMITTER_EMAIL = getattr(settings, 'WALIKI_COMMITTER_EMAIL', 'waliki@waliki.pythonanywhere.com')

WALIKI_COMMITTER_NAME = getattr(settings, 'WALIKI_COMMITTER_NAME', 'Waliki')

WALIKI_CACHE_TIMEOUT = getattr(settings, 'WALIKI_CACHE_TIMEOUT', 60*60*24)

WALIKI_ATTACHMENTS_DIR = getattr(settings, 'WALIKI_ATTACHMENTS_DIR', None)  or _get_default_data_dir('waliki_attachments', False)

WALIKI_UPLOAD_TO_PATTERN = '%(slug)s/%(filename)s'

WALIKI_RST_DIRECTIVES = getattr(settings, 'WALIKI_RST_DIRECTIVES', ['waliki.directives.embed'])

WALIKI_RST_TRANSFORMS = getattr(settings, 'WALIKI_RST_TRANSFORMS', ['waliki.directives.transforms.Emojis'])

WALIKI_USE_MATHJAX = getattr(settings, 'WALIKI_USE_MATHJAX', False)

WALIKI_BREADCRUMBS = getattr(settings, 'WALIKI_BREADCRUMBS', False)


def WALIKI_UPLOAD_TO(instance, filename):
    return os.path.join(WALIKI_ATTACHMENTS_DIR,
                        WALIKI_UPLOAD_TO_PATTERN % {'slug': instance.page.slug,
                                                       'page_id': getattr(instance.page, 'id', ''),
                                                       'filename': filename,
                                                       'filename_extension': os.path.splitext(filename)[1]})

for mod in WALIKI_RST_DIRECTIVES:
    register_directive = import_string(mod + '.register_directive')
    register_directive()
